---
title: "Coverage Calculation"
author: "QiongJia"
date: "`r Sys.Date()`"
output: 
  html_document:
    highlight: pygments
    toc: true
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Design  

To calculate the average coverage(sequencing depth) for the given sample BAM file, two approach are presented below.  

1. The first one is based on the coverage calculation equation:   

$$C = \frac{LN}{G}$$  

where    
 - $C$ is coverage.  
 - $G$ is the haploid genome length.  
 - $L$ is the read length in the sequencing.  
 - $N$ is the number of reads.  
 
   The haploid reference genome (GRCh38) size is $3,099,734,149$ bases ([Reference](https://www.ncbi.nlm.nih.gov/grc/human/data)). The read length and number can be obtained from ***.bam.bas*** file. 

2. The second one is more precise. The ***samtool depth*** can be used to calculate the coverage at each genomic position and the average coverage of the given BAM file.  Same as ***bedtools genomecov***. 

3. At the end, one another tools ***mosdepth*** is presented briefly in coverage calculation.   

## Download BAM file 

```{bash download, eval=FALSE}
# make working directory
cd ~
mkdir TakeHomeFulgent
cd TakeHomeFulgent

# download cram file
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/data/CEU/NA12878/alignment/NA12878.alt_bwamem_GRCh38DH.20150718.CEU.low_coverage.cram

# download cram index file
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/data/CEU/NA12878/alignment/NA12878.alt_bwamem_GRCh38DH.20150718.CEU.low_coverage.cram.crai

# download .bam.bas file
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/data/CEU/NA12878/alignment/NA12878.alt_bwamem_GRCh38DH.20150718.CEU.low_coverage.bam.bas

```

```{bash define, eval=FALSE}
# define work variable
cram=NA12878.alt_bwamem_GRCh38DH.20150718.CEU.low_coverage.cram
bas=NA12878.alt_bwamem_GRCh38DH.20150718.CEU.low_coverage.bam.bas
```

## Quick estimate 

```{bash quick, eval=FALSE}
# reference genome size
G=3099734149
# read length equals total bases number / total reads number
L=$(awk 'NR==2 {print $9 / $11}' "$bas")
# reads number equals total reads number - duplicate reads number
N=$(awk 'NR==2 {print $11-$21}' "$bas")
# coverage calculation
C=$(echo "$N*$L/$G" | bc -l)
printf "The average coverage is: %.2f\n", "$C"
```

The average coverage is: **5.77**  

## Coverage calculation for each position 

### samtools  

**Step1**: calculate the coverage at each genomic position. 

```{bash samtools depth, eval=FALSE}
samtools depth -a $cram > NA12878_cram_coverage.txt
```

Overlook the coverage output.   

```{bash samtools overlook, eval=FALSE}
head -n 5 NA12878_cram_coverage.txt
```
  
```{r samtools overlookr, echo=FALSE}
head <- read.delim("NA12878_cram_coverage_head.txt", header = F, sep = "\t")
head
```
  
Each line represents a genomic position. Three columns are included int he coverage output:   

- Chromosome;  
- Position;  
- Reads covered this position.   

**Step2**: calculate the average coverage. 

```{bash samtools average, eval=FALSE}
awk '{sum+=$3} END { print "Average coverage from samtools = ",sum/NR}' NA12878_cram_coverage.txt
```

Average coverage from samtools = **5.14261**

### bedtools  

***bedtools genomecov -d*** also reports the genome coverage per base as below:  

```{bash bedtools, eval=FALSE}
# To use -ibam flag in bedtools genomecov, the bam file is needed to be sorted by position
samtools sort $cram | bedtools genomecov -ibam stdin -d > NA12878_cram_genomecov.txt
```

Then the average coverage would be: 

```{bash bedtools_average, eval=FALSE}
awk '{sum+=$3} END { print "Average coverage from bedtools = ",sum/NR}' NA12878_cram_genomecov.txt
```

Average coverage from bedtools = **5.59374**


## Other avaialble tools 

***mosdepth*** can report coverage for both per-base and summary result at the same time. It can take **BAM** file directly, but it needs reference genome file with **CRAM** file as input. 

```{bash mosdepth, eval=FALSE}
mosdepth NA12878 $cram
```

The file ended with ***.mosdepth.summary.txt*** contain the average coverage result.

Besides, It can also report coverage based on the user defined region by using ***--by <bed|window>***. 

## Runing time comparison  

- *samtools depth* :  1176s;
- *bedtools genomecov* : 32812s;

```{r session, echo=FALSE}
sessionInfo()
```